Remove circular deps

dsmilkov opened the pull request:
Remove all circular dependencies and add a lint check.

Circular dependencies prohibit us to properly sync the library internally at Google. This PR has no changes in functionality. Just moving code around.

An overview of the changes:
- `tensor.ts` becomes a lightweight module with almost no dependencies. It doesn't depend on any environment/engine/ops anymore. Instead the environment registers an op handler and a tensor tracker, and the Tensor class delegates chain ops and registration to those handlers (see `tensor.ts`)
- Combined `environment.ts` and `engine.ts` in one file as they were already deeply coupled
- Split `types.ts` into `types.ts` and `tensor_types.ts` (the latter depends on tensor)
- Split `util.ts` into `util.ts` and `tensor_util.ts` (the latter depends on tensor)
- Split `array_ops.ts` into `array_ops.ts` and `tensor_ops.ts` (the latter doesn't call into ENV.engine, the former calls ENV.engine.runKernel)
- Other minor splits (e.g. `relu_ops.ts` separated from `unary_ops.ts` because of circular dep, etc...)

DEV

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-core/1117)
<!-- Reviewable:end -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 31 files at r1, 1 of 48 files at r3.
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/environment.ts, line 660 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-LFbxyhrCDFHb3LFsE39:-LFbxyhrCDFHb3LFsE3A:b-fvub57) ([raw file](https://github.com/tensorflow/tfjs-core/blob/344d81a896df40ba149e10fa627c67fe6c35e6cc/src/environment.ts#L660)):*
> ```XML
> const TEST_EPSILON_FLOAT32_DISABLED = 1e-1;
> 
> function hasExtension(gl: WebGLRenderingContext, extensionName: string) {
> ```

this file is enormous now, it was already growing out of hand.

There are a bunch of functions in this file that are leaves that do feature testing (all of the WebGL functions from here down), can you move them to a new file, maybe environment_util.ts?

---

*[src/tensor.ts, line 1261 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-LFbz42OBnzR3IMGDotk:-LFbyswb9VqaQ4GHjY0l:b-9a9aas) ([raw file](https://github.com/tensorflow/tfjs-core/blob/344d81a896df40ba149e10fa627c67fe6c35e6cc/src/tensor.ts#L1261)):*
> ```XML
> }
> 
> // Maximum number of values before we decide to show ellipsis.
> ```

this file is also huge, could you move these functions out to tensor_util.ts (to remove dependency on Tensor just pass vals, shape, and strides directly)

---

*[src/tensor_types.ts, line 50 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-LFc-WBT3j_56gcDRgpw:-LFc-WBT3j_56gcDRgpx:bblcpwp) ([raw file](https://github.com/tensorflow/tfjs-core/blob/344d81a896df40ba149e10fa627c67fe6c35e6cc/src/tensor_types.ts#L50)):*
> ```XML
> }
> 
> export interface ModelPredictConfig {
> ```

how about a model_types.ts file for these two interfaces (predict and inference model)?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-:-LFc3Sb14jOdYT5L6oCa:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Reviewed 31 of 31 files at r1, 3 of 3 files at r2, 48 of 48 files at r3, 27 of 27 files at r4, 5 of 7 files at r5.
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/environment.ts, line 660 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-LFbxyhrCDFHb3LFsE39:-LFd0BOE6ab8kuZAkGHL:b-xua2g1) ([raw file](https://github.com/tensorflow/tfjs-core/blob/344d81a896df40ba149e10fa627c67fe6c35e6cc/src/environment.ts#L660)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

this file is enormous now, it was already growing out of hand.

There are a bunch of functions in this file that are leaves that do feature testing (all of the WebGL functions from here down), can you move them to a new file, maybe environment_util.ts?
</blockquote></details>

Done! I split `environment.ts` into 3 files:
- `environment.ts` 406 LOC
- `engine.ts` 474 LOC
- `environment_util.ts` 284 LOC

---

*[src/tensor.ts, line 1261 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-LFbz42OBnzR3IMGDotk:-LFcWtneCc-K_IhqPFB0:bv8j4g7) ([raw file](https://github.com/tensorflow/tfjs-core/blob/344d81a896df40ba149e10fa627c67fe6c35e6cc/src/tensor.ts#L1261)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

this file is also huge, could you move these functions out to tensor_util.ts (to remove dependency on Tensor just pass vals, shape, and strides directly)
</blockquote></details>

Done. Moved to `tensor_format.ts` since `tensor_util.ts` depends on Tensor since it has a bunch of utils that work with Tensors

---

*[src/tensor_types.ts, line 50 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1117#-LFc-WBT3j_56gcDRgpw:-LFcYFfeELrgf3N3y_a9:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/344d81a896df40ba149e10fa627c67fe6c35e6cc/src/tensor_types.ts#L50)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

how about a model_types.ts file for these two interfaces (predict and inference model)?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1117)*
<!-- Sent from Reviewable.io -->
